File: framework/ui-elements/src/displays/components/BaseDigitalSignature.js
import React from "react";
import { withStyles } from '@material-ui/core/styles';
import AbstractBaseDigitalSignature from "../../../gen/displays/components/AbstractBaseDigitalSignature";
import BaseDigitalSignatureNotifier from "../../../gen/displays/notifiers/BaseDigitalSignatureNotifier";
import BaseDigitalSignatureRequester from "../../../gen/displays/requesters/BaseDigitalSignatureRequester";
import DisplayFactory from 'alexandria-ui-elements/src/displays/DisplayFactory';
import { withSnackbar } from 'notistack';

const styles = theme => ({});

export default class BaseDigitalSignature extends AbstractBaseDigitalSignature {

	constructor(props) {
		super(props);
		this.notifier = new BaseDigitalSignatureNotifier(this);
		this.requester = new BaseDigitalSignatureRequester(this);
		this.state = {
		    ...this.state,
		    data: null,
		    isDocument: false,
		    signing: false,
		    readonly: false,
		};
	};

    text = (content) => {
        this.setState({ data: content, isDocument: false });
    };

    sign = () => {
    };

    document = (content) => {
        this.setState({ data: content, isDocument: true });
    };

    refreshReadonly = (value) => {
        this.setState({readonly: value});
    };

}

File: framework/ui/src/io/intino/alexandria/ui/displays/components/BaseDigitalSignature.java
package io.intino.alexandria.ui.displays.components;

import io.intino.alexandria.Base64;
import io.intino.alexandria.core.Box;
import io.intino.alexandria.schemas.DigitalSignatureFailure;
import io.intino.alexandria.schemas.DigitalSignatureSuccess;
import io.intino.alexandria.ui.displays.events.SignErrorEvent;
import io.intino.alexandria.ui.displays.events.SignErrorListener;
import io.intino.alexandria.ui.displays.events.SignEvent;
import io.intino.alexandria.ui.displays.events.SignListener;
import io.intino.alexandria.ui.displays.notifiers.BaseDigitalSignatureNotifier;
import io.intino.icod.core.SignatureInfo;

import java.nio.charset.StandardCharsets;

public abstract class BaseDigitalSignature<DN extends BaseDigitalSignatureNotifier, B extends Box> extends AbstractBaseDigitalSignature<DN, B> {
    private SignListener signListener;
    private SignErrorListener errorListener;
    private String text;
    public boolean readonly = false;

    public BaseDigitalSignature(B box) {
        super(box);
    }

    @Override
    public void didMount() {
        super.didMount();
        notifier.refreshReadonly(readonly);
    }

    @Override
    public void refresh() {
        super.refresh();
        if (text != null) notifier.text(text);
    }

    public BaseDigitalSignature<DN, B> readonly(boolean readonly) {
        this.readonly = readonly;
        notifier.refreshReadonly(readonly);
        return this;
    }

    public void text(String content) {
        text = Base64.encode(content.getBytes(StandardCharsets.UTF_8));
        notifier.text(text);
    }

    public void sign() {
        notifier.sign();
    }

    public void success(DigitalSignatureSuccess success) {
        signListener.accept(new SignEvent(this, success.signature(), success.certificate(), info(success.signature())));
    }

    public void failure(DigitalSignatureFailure failure) {
        errorListener.accept(new SignErrorEvent(this, failure.code(), failure.message()));
    }

    public BaseDigitalSignature<DN, B> onSign(SignListener listener) {
        this.signListener = listener;
        return this;
    }

    public BaseDigitalSignature<DN, B> onError(SignErrorListener listener) {
        this.errorListener = listener;
        return this;
    }

    protected abstract SignatureInfo info(String signature);

}